### PR TITLE
Kubernetes secret encryption integration. Adds support for 3 providers: AWS KMS, IBMCloud KMS, and AESCBC. IBMCloud KMS is critical to IBMCloud's Fedramp strategy

### PR DESCRIPTION
**Pull request by relyt0925**
Google doc outlining customer requirements in the various hypershift environments:
https://docs.google.com/document/d/1ondCNNi13awqC37neB5gc5h8Wa9qYmSpCbWDaWspvxE

### CODE REVIEW

**Review by csrwng — COMMENTED** [review]
@relyt0925 some comments below. In general looks good to me.

I tried out both aescbc and aws encryption. The one thing I found with aescbc is that updating HostedCluster to rotate the keys (move active -> backup, add new active) doesn't result in kube-apiserver getting restarted so the change is not picked up. Would it be possible to add a hash of the encryption config as an annotation to the deployment? (like https://github.com/openshift/hypershift/blob/main/control-plane-operator/controllers/hostedcontrolplane/kas/deployment.go#L114)

File: api/v1alpha1/hostedcluster_types.go [hunk 1]
```diff
@@ -406,6 +426,177 @@ type EtcdTLSConfig struct {
...
+	// Region contains the AWS region
+	Region string `json:"region"`
+	// ActiveKey defines the active key used to encrypt new secrets
+	ActiveKey *AWSKMSKeyEntry `json:"activeKey"`
+	// BackupKey defines the old key during the rotation process so previously created
+	// secrets can continue to be decrypted until they are all re-encrypted with the active key.
+	// +optional
+	BackupKey *AWSKMSKeyEntry `json:"backupKey,omitempty"`
+	// Auth defines metadata about the management of credentials used to interact with AWS KMS
+	Auth *AWSKMSAuthSpec `json:"auth"`
```
[the remark this answers is not in the document]
Comment (relyt0925) [reply api/v1alpha1/hostedcluster_types.go:556]: done

File: api/v1alpha1/hostedcluster_types.go [hunk 2]
```diff
@@ -406,6 +426,177 @@ type EtcdTLSConfig struct {
...
+	// +optional
+	AWS *AWSKMSSpec `json:"aws,omitempty"`
+}
+
+// IBMCloudKMSSpec defines metadata for the IBM Cloud KMS encryption strategy
+type IBMCloudKMSSpec struct {
+	// Region is the IBM Cloud region
+	Region string `json:"region"`
+	// Auth defines metadata for how authentication is done with IBM Cloud KMS
+	Auth *IBMCloudKMSAuthSpec `json:"auth"`
```
[the remark this answers is not in the document]
Comment (relyt0925) [reply api/v1alpha1/hostedcluster_types.go:485]: changed

File: api/v1alpha1/hosted_controlplane.go [hunk 3]
```diff
@@ -85,6 +85,10 @@ type HostedControlPlaneSpec struct {
 	// ImageContentSources lists sources/repositories for the release-image content.
 	// +optional
 	ImageContentSources []ImageContentSource `json:"imageContentSources,omitempty"`
+
+	// SecretEncryption contains metadata about the kubernetes secret encryption strategy being used for the
+	// cluster when applicable.
```
[the remark this answers is not in the document]
Comment (relyt0925) [reply api/v1alpha1/hosted_controlplane.go:95]: added

File: api/v1alpha1/hostedcluster_types.go [hunk 4]
```diff
@@ -406,6 +426,177 @@ type EtcdTLSConfig struct {
...
+type AWSKMSKeyEntry struct {
+	// ARN is the Amazon Resource Name for the encryption key
+	// +kubebuilder:validation:Pattern=`^arn:`
+	ARN string `json:"arn"`
+}
+
+// AESCBCSpec defines metadata about the AESCBC secret encryption strategy
+type AESCBCSpec struct {
+	// ActiveKey defines the active key used to encrypt new secrets
+	ActiveKey *corev1.LocalObjectReference `json:"activeKey"`
```
[the remark this answers is not in the document]
Comment (relyt0925) [reply api/v1alpha1/hostedcluster_types.go:577]: done

File: control-plane-operator/controllers/hostedcontrolplane/kas/aescbc.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+		},
+		Resources: []v1.ResourceConfiguration{
+			{
+				Resources: []string{"secrets"},
+				Providers: providerConfiguration,
+			},
+		},
+	}
+	bufferInstance := bytes.NewBuffer([]byte{})
+	err = api.YamlSerializer.Encode(&encryptionConfig, bufferInstance)
```
[the remark this answers is not in the document]
Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/kas/aescbc.go:62]: done

File: api/v1alpha1/hostedcluster_types.go [hunk 6]
```diff
@@ -408,6 +428,159 @@ type EtcdTLSConfig struct {
...
+	// AESCBC uses AES-CBC with PKCS#7 padding to do secret encryption
+	AESCBC SecretEncryptionType = "aescbc"
+)
+
+// SecretEncryptionSpec contains metadata about the kubernetes secret encryption strategy being used for the
+// cluster when applicable.
+type SecretEncryptionSpec struct {
+	// SecretEncryptionType defines the type of kube secret encryption being used
+	// +unionDiscriminator
+	SecretEncryptionType SecretEncryptionType `json:"type"`
```
[the remark this answers is not in the document]
Comment (relyt0925) [reply api/v1alpha1/hostedcluster_types.go]: Done

File: control-plane-operator/controllers/hostedcontrolplane/kas/ibmcloud_kms.go [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+	"k8s.io/utils/pointer"
+	"sort"
+	"strconv"
+	"time"
+)
+
+var (
+	ibmCloudKMSVolumeMounts = util.PodVolumeMounts{
+		kasContainerMain().Name: {
+			kasVolumeKMSSocket().Name: "/tmp",
```
[the remark this answers is not in the document]
Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/kas/ibmcloud_kms.go:23]: for ibmcloud we need to use tmp currently (the code is setup with some paths that assume tmp) in the future if that gets configurable we can adjust from tmp for our provider.

I would like to do that if applicable separate from this pr

File: control-plane-operator/controllers/hostedcontrolplane/kas/aws_kms.go [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+	backupAWSKMSUnixSocket         = "unix:///tmp/awskmsbackup.sock"
+	backupAWSKMSUnixSocketFilePath = "/tmp/awskmsbackup.sock"
+	backupAWSKMSHealthPort         = 8081
+	awsKeyNamePrefix               = "awskmskey"
+)
+
+var (
+	awsKMSVolumeMounts = util.PodVolumeMounts{
+		kasContainerMain().Name: {
+			kasVolumeKMSSocket().Name: "/tmp",
```
[the remark this answers is not in the document]
Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/kas/aws_kms.go:30]: for ibmcloud we need to use tmp currently (the code is setup with some paths that assume tmp) in the future if that gets configurable we can adjust from tmp for our provider.

I would like to do that if applicable separate from this pr

[the remark this answers is not in the document]
Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/kas/aws_kms.go:30]: Will adjust for aws provider

**Pull request merged**